pkg/stewardlabels/propagate.go

```diff
@@ -0,0 +1,93 @@
+package stewardlabels
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+/*
+propagate propagates specified labels from `sourceObj` to `destObj`,
+while values can be enforced for individual label keys.
+In case of a value conflict an error is returned and `destObj` is
+not modified.
+
+The set of labels to propagate is defined by the keys of `labelSpec`.
+A label value is enforced if the corresponding key in `labelSpec` is
+associated with a non-empty string value (empty values cannot be
+enforced).
+
+A value conflict exists for a label key if
+	a) the label value is enforced and the existing value at `sourceObj`
+	   differs from the enforced value, or
+	b) the label value is enforced and the exiting value at `destObj`
+	   differs from the enforced value, or
+    c) the existing values at `sourceObj` and `destObject` differ.
+The empty string value is NOT treated specially, e.g. there's a conflict
+if `destObj` has a label set with value "foo" but `sourceObj` has the
+same label set with an empty string value.
+
+If a value is enforced for a key, the label gets set on `destObj` even
+if `sourceObj` does not have that label set.
+If a value is not enforced and `sourceObj` has that label set, `destObj`
+get the same label set, except there is a value confict which leads to an
+error.
+If a value is not enforced and `sourceObj` does NOT have that label set,
+the respective label at `destObj` remains unchanged, i.e. will not be
+created, deleted or modified.
+*/
+func propagate(destObj metav1.Object, sourceObj metav1.Object, labelSpec map[string]string) error {
+	destLabels := destObj.GetLabels()
+	sourceLabels := sourceObj.GetLabels()
+
+	// don't modify labels of dest object until propagation finished without errors
+	propagatedLabels := make(map[string]string)
+
+	// fail if source has any value conflict with enforced value
+	for k, v := range labelSpec {
+		if v != "" { // value enforced
+			sourceValue, found := sourceLabels[k]
+			if found && sourceValue != v {
+				return fmt.Errorf(
+					"value conflict: source object label %q has value %q but %q is expected",
+					k, sourceValue, v,
+				)
+			}
+		}
+	}
+
+	// propagate
+	for k, v := range labelSpec {
+		if v == "" { // value not enforced
+			var foundAtSource bool
+			v, foundAtSource = sourceLabels[k]
+			if !foundAtSource {
+				continue
+			}
+		}
+		destValue, found := destLabels[k]
+		if found {
+			if destValue != v {
+				return fmt.Errorf(
+					"value conflict: destination object label %q has existing value %q but %q is expected",
+					k, destValue, v,
+				)
+			}
+		} else {
+			propagatedLabels[k] = v
+		}
+	}
```

Comment:
```suggestion
	// fail if source has any value conflict with enforced value
	for k, v := range labelSpec {
	        sourceValue, foundAtSource := sourceLabels[k]
		if v == ""  { // value not enforced, take it from source if exists
			if !foundAtSource {
				continue
			} else {
			        v = sourceValue
			}
		} else { // value enforced
			if foundAtSource && sourceValue != v {
				return fmt.Errorf(
					"value conflict: source object label %q has value %q but %q is expected",
					k, sourceValue, v,
				)
			}
		}
		destValue, found := destLabels[k]
		if found {
			if destValue != v {
				return fmt.Errorf(
					"value conflict: destination object label %q has existing value %q but %q is expected",
					k, destValue, v,
				)
			}
		} else {
			propagatedLabels[k] = v
		}
	}
```

Reply:
This avoids to loop twice and INHO is still readable.

Reply:
My intention was to _first_ check the source object for value conflicts and possibly report them _before_ looking at the destination object and possibly report a value conflict there.

Reply:
I think it's better to first check the source object completely. Therefore I would keep it as it is.

Nonetheless I've applied some minor changes in 8e7971a.